### DISCUSSION TITLE
What is the best way to get the address of a contract deployed with deployContract method?

### DISCUSSION THREAD

**Discussion by 0xFloop** [wevm/viem#321] [Question] [answered]
Looking to immediately have and use the contract address of a newly deployed contract. Is the only way to do this manually calculating it prior to deployment? Thanks for any help!

**Comment by tmm** [comment, marked as answer]
Check out [`getContractAddress`](https://viem.sh/docs/utilities/getContractAddress.html).

**Reply by anthonygood** [reply]
Where should one get the `from` and `nonce` values? The example includes two magic values:
```ts
getContractAddress({ 
  from: '0xc961145a54C96E3aE9bAA048c4F4D6b04C13916b',
  nonce: 69420n
})
```

**Reply by jxom** [reply]
You should get it from the deployment transaction.

**Comment by 0xFloop** [comment]
The only reason I thought using getContractAddress was a hassle is that I couldn't find a getNonce method but i'm realizing i'm an idiot and its the getTransactionCount method. I got it working now, thanks for the help :)

**Comment by AlphaDio** [comment]
This is still a hassle and very clunky when compared to the workflow of ethers. I don't see why the object returned after deploying could not be enriched with the deployed Contract info.

**Reply by tmm** [reply]
Appreciate the feedback! In general, we try to develop for [clarity over cleverness](http://www.catb.org/~esr/writings/taoup/html/ch01s06.html) and keep things modular. Once you start adding extra internal magic to functions, things tend to get bloated.

It's a bit extra work to call one function (especially when you are used to using another library), but it should make things more stable and robust in the long term.

**Reply by jxom** [reply]
Including the contract address also may confuse consumers that the contract may already be deployed (whereas it may not be). This is something we can only really convey through docs. Having these functions separated means that the consumer can reason about the functionality, and understand how contracts are actually deployed & how addresses are derived on Ethereum.

We have designed the modules so you can easily compose them together – ie. you can create your own opinionated `deployContract` function by composing viem’s `deployContract` & `getContractAddress` & whatever else you want in there.

**Comment by hungdoansy** [comment]
Hi, if you're using `deployContract`, you can wait for the tx receipt and here's the contract address: `txReceipt.contractAddress`
Source: https://github.com/wevm/viem/blob/viem%402.10.3/src/types/transaction.ts#L44
